### ISSUE TITLE
multi machine train seem efficient but low accuracy

### ISSUE THREAD

**Issue by GekTo** [apache/incubator-mxnet#2166]
Hi all,
I am testing the distributed train over multi-machine mxnet. I used the train_mnist.py and test with lenet. And I set one machine as dispatcher， and a half of the rest machine as workers and the rest as server.
However, even though the time used for train one epoch is linear decrease as the machine I used, the train accuracy and validation accuracy isn't increase with the train epoch any more.
I have checked that the ps-lite is working fine, cause when set PS_VERBOSE=1, connection info is right, set PS_VERBOSE=2, lots of communication log is print.
So, How to explain those low accuracy and no increase of it?

the following is some log info parsed from tools/parselog.py
1 node:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.766714 | 0.904172 | 238.0 |
|  2 | 0.853887 | 0.889674 | 229.8 |
|  3 | 0.921867 | 0.887795 | 228.5 |
|  4 | 0.935588 | 0.952324 | 230.6 |
|  5 | 0.934608 | 0.932267 | 229.3 |
|  6 | 0.938147 | 0.933068 | 230.7 |
|  7 | 0.938343 | 0.940856 | 228.8 |
|  8 | 0.937893 | 0.945137 | 226.7 |
|  9 | 0.933089 | 0.941256 | 230.7 |
| 10 | 0.879812 | 0.903922 | 227.7 |
5 nodes:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.650182 | 0.648287 | 119.8 |
|  2 | 0.649210 | 0.651492 | 120.1 |
|  3 | 0.650011 | 0.648387 | 121.1 |
|  4 | 0.653091 | 0.651267 | 121.8 |
|  5 | 0.695041 | 0.783979 | 121.3 |
|  6 | 0.694958 | 0.649164 | 121.6 |
|  7 | 0.666900 | 0.650115 | 121.2 |
|  8 | 0.649944 | 0.649714 | 121.2 |
|  9 | 0.649923 | 0.652294 | 120.7 |
| 10 | 0.655019 | 0.652794 | 120.3 |
8 nodes:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.647549 | 0.647949 | 63.5 |
|  2 | 0.649656 | 0.648951 | 62.5 |
|  3 | 0.649381 | 0.655633 | 62.2 |
|  4 | 0.650186 | 0.646125 | 62.5 |
|  5 | 0.651363 | 0.654194 | 62.6 |
|  6 | 0.676048 | 0.695826 | 62.3 |
|  7 | 0.653946 | 0.646716 | 62.1 |
|  8 | 0.659827 | 0.651650 | 62.7 |
|  9 | 0.667860 | 0.668586 | 62.9 |
| 10 | 0.708588 | 0.683080 | 62.9 |
10 nodes:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.648937 | 0.647786 | 51.3 |
|  2 | 0.649609 | 0.648516 | 50.6 |
|  3 | 0.649013 | 0.647604 | 50.0 |
|  4 | 0.648921 | 0.646510 | 49.2 |
|  5 | 0.649273 | 0.652031 | 49.5 |
|  6 | 0.651932 | 0.652240 | 49.2 |
|  7 | 0.650055 | 0.644635 | 49.6 |
|  8 | 0.651932 | 0.651458 | 49.4 |
|  9 | 0.681527 | 0.651875 | 49.9 |
| 10 | 0.666961 | 0.682656 | 50.0 |
20 nodes:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.648939 | 0.645368 | 25.5 |
|  2 | 0.649168 | 0.645703 | 25.1 |
|  3 | 0.649907 | 0.646708 | 25.1 |
|  4 | 0.649588 | 0.648884 | 24.9 |
|  5 | 0.650000 | 0.648493 | 25.0 |
|  6 | 0.650255 | 0.645089 | 24.6 |
|  7 | 0.648964 | 0.646233 | 24.8 |
|  8 | 0.649011 | 0.644029 | 24.4 |
|  9 | 0.649231 | 0.645954 | 25.0 |
| 10 | 0.650972 | 0.645368 | 24.8 |
50 nodes:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.648767 | 0.648776 | 10.3 |
|  2 | 0.648980 | 0.647396 | 9.9 |
|  3 | 0.649384 | 0.649922 | 9.9 |
|  4 | 0.649948 | 0.644609 | 9.8 |
|  5 | 0.649427 | 0.646901 | 9.9 |
|  6 | 0.648490 | 0.646667 | 10.0 |
|  7 | 0.647743 | 0.646901 | 9.9 |
|  8 | 0.648655 | 0.648333 | 9.9 |
|  9 | 0.648329 | 0.648047 | 9.8 |
| 10 | 0.649002 | 0.649479 | 9.9 |
100 nodes:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.647895 | 0.647227 | 5.6 |
|  2 | 0.647717 | 0.655469 | 5.0 |
|  3 | 0.650195 | 0.648984 | 4.9 |
|  4 | 0.649154 | 0.649805 | 4.9 |
|  5 | 0.649839 | 0.644766 | 5.0 |
|  6 | 0.648190 | 0.647461 | 4.9 |
|  7 | 0.648954 | 0.648906 | 4.9 |
|  8 | 0.648411 | 0.653633 | 5.0 |
|  9 | 0.649926 | 0.645547 | 4.9 |
| 10 | 0.649115 | 0.649414 | 4.9 |

BTW,  the cpu-only version is quite slow, anyway to speed up?

**Comment by piiswrong** [comment]
Are you using async or sync

**Comment by GekTo** [comment]
I used sync

**Comment by piiswrong** [comment]
I'm guessing it's due to batch size or learning rate

**Comment by GekTo** [comment]
Thanks, after change the batch size to 64 or 32 which is 128 originally used for one node train, the accuracy is increased to 88% now, with 10 nodes, 8 workers, 1 dispatcher, 1 server. 
I'm confused, it that mean we need to decrease the batch size where distributed training ?
If so, is there any best practice on how to decrease the batch size or learning rate?
And still, I not so sure how much server and worker to used for better performance?
And it save model on every nodes, which to used or they are the same?

So many question, I would be grateful if you could provide any help !
@piiswrong

**Comment by piiswrong** [comment]
divide by the number of nodes would give you exactly the same result (i think @mli could you confirm?) with sync. But it may be slow. 

Hyper parameter tuning is in general messy. Do it enough times and you'll develop intuition.

**Comment by mli** [comment]
the effective batch size for sync is num_node * batch_size. as for mnist, a large minibatch size (i guess >256) affect the convergence a lot. but for large dataset, such as imagenet, it is less sensitive to the batch size.

**Comment by GekTo** [comment]
Thanks, I get the same accuracy with batch size 16 for 8 nodes training as batch size 128 for 1nodes training.
1 node:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
| 1 | 0.766714 | 0.904172 | 238.0 |
| 2 | 0.853887 | 0.889674 | 229.8 |
| 3 | 0.921867 | 0.887795 | 228.5 |
| 4 | 0.935588 | 0.952324 | 230.6 |
| 5 | 0.934608 | 0.932267 | 229.3 |
| 6 | 0.938147 | 0.933068 | 230.7 |
| 7 | 0.938343 | 0.940856 | 228.8 |
| 8 | 0.937893 | 0.945137 | 226.7 |
| 9 | 0.933089 | 0.941256 | 230.7 |
| 10 | 0.879812 | 0.903922 | 227.7 |
8 nodes:
| epoch | train-accuracy | valid-accuracy | time |
| --- | --- | --- | --- |
|  1 | 0.688376 | 0.858774 | 48.7 |
|  2 | 0.885074 | 0.930639 | 49.7 |
|  3 | 0.916800 | 0.927259 | 50.6 |
|  4 | 0.930731 | 0.946990 | 50.3 |
|  5 | 0.917748 | 0.899790 | 49.3 |
|  6 | 0.903274 | 0.918570 | 49.1 |
|  7 | 0.922764 | 0.935397 | 50.4 |
|  8 | 0.930051 | 0.941782 | 50.8 |
|  9 | 0.936661 | 0.940905 | 50.5 |
| 10 | 0.936465 | 0.944837 | 49.0 |

Seems that small batch size have bad effect on the efficiency.
I would test with imageNet to get better comprehension of those issue.

Still have some question:
1、 "the effective batch size for sync is num_node * batch_size", how about the effective batch size for async mode?
2、 how much server and worker to used for better performance ?
I think that the parameter server is used for release the communication  pressure of many workers' pulling and pushing to one server, but too much server would cause the sync communication to become huge, is that right? It seems that the default setting is the same number of worker and server， is that the best practice setting?
I test the same number of worker with different number of server, there isn't obvious change of the efficiency.
3、 it save model on every nodes, which to used or they are the same?
This is quite a dump question but my cluster environment is sharing one file system, and the saved model is all named the same and keep covered by the latest node model, so I can't check it myself.

@mli @piiswrong thanks a lots.

**Comment by mli** [comment]
1. refer to http://mxnet.dmlc.ml/en/latest/system/multi_node.html
2. the number of servers doesn't matter too much
3. picking a random one is often fine

**Comment by GekTo** [comment]
Thank you for answer my question. 
Seem that async mode works better normally.
But it seems that when training in async on my cluster, the program just hang after training all the epochs, instead of log info from van.cc tell workers and servers stoping and terminated when training in sync mode. So I have to kill it by my own. Not a serious problem though.

**Comment by mli** [comment]
I thought the last hanging problem was fixed by adding a timeout for
termination. But it seems appears again after upgrading Zmq
